src/front/spv/mod.rs

```diff
@@ -1984,6 +1990,90 @@ impl<I: Iterator<Item = u32>> Parser<I> {
                 Op::Fwidth | Op::FwidthFine | Op::FwidthCoarse => {
                     self.parse_expr_derivative(expressions, crate::DerivativeAxis::Width)?;
                 }
+                Op::ArrayLength => {
+                    inst.expect(5)?;
+                    let result_type_id = self.next()?;
+                    let result_id = self.next()?;
+                    let structure_id = self.next()?;
+                    let member_index = self.next()?;
+
+                    // Validate that all the types are correct.
+                    let result_type = self.lookup_type.lookup(result_type_id)?;
+                    match type_arena[result_type.handle].inner {
+                        crate::TypeInner::Scalar {
+                            kind: crate::ScalarKind::Uint,
+                            width: 4,
+                        } => {}
+                        _ => return Err(Error::InvalidParameter(Op::ArrayLength)),
+                    }
+                    let structure = self.lookup_expression.lookup(structure_id)?;
+                    let structure_type = self.lookup_type.lookup(structure.type_id)?;
+                    match type_arena[structure_type.handle].inner {
+                        crate::TypeInner::Pointer { base, .. } => {
+                            if let crate::TypeInner::Struct { ref members, .. } = type_arena[base].inner {
+                                if let Some(last) = members.last() {
+                                    match type_arena[last.ty].inner {
+                                        crate::TypeInner::Array { size: crate::ArraySize::Dynamic, .. } => if members.len() - 1 == member_index as _ {},
```

Comment:
I don't think we need to be that detailed here. Let the validation catch what it can. So here, let's instead use the member index directly:
```rust
match type_arena[members[member_index as u32].ty].inner {
}
```

Reply:
This actually touches on something I'm wondering: should naga never panic, no matter what input it gets?

Reply:
Naga backends are generally free to panic on the IR that's not valid. However, even there we'd want to issue `Error::Validation` by the backend, because this condition may be a result of some miscommunication between the validator and the backend, so these conditions may still arise, and we still don't want to panic there.

Naga frontends, however, shouldn't panic, because we have no guarantee or expectation that there was any validation before us. That is to say, the frontends aren't expected to fully validate the source files, it's not our job.